File: backend/TreasureHunt/php/TreasureHunt.php
<?php

class TreasureHunt
{
    private $record;   //記録
    private $gold;     //所持ゴールド
    private $position; //現在地
    private $turn;     //ターン数

    function __construct() {
        $this->record   = "";
        $this->gold     = 0;
        $this->position = 0;
        $this->turn     = 0;
    }
    public function execute($number, $arrInput): string
    {
        for ($i = 0; $i < strlen($arrInput); $i++) {
            //ターンを進める
            $this->turn++;
            //現在地の島クラスのインスタンスを取得する
            $land = LandControllerFactory::getController($this->position);
            //ダイスの出目を取得する
            $dice = $land->getDice($arrInput, $i);
            //記録内容を更新する
            $prefix = empty($this->record) ? "" : ", ";
            $this->record .= $land->getCurrentTurnRecord($dice, $prefix);
            //次のターンの現在地を取得する
            $this->position = $land->getNextPosition($dice, $this->gold);

            //ゴールに着いたら残りターン数があってもループを抜ける
            if ($this->position == Land::LandGoal) {
                $this->record .= ", G";
                break;
            }
        }

        return $number.', "'.$arrInput.'", '.$this->turn.", ".$this->gold.", ".'"'.$this->record.'"';
    }
}

class LandControllerFactory
{
    static public function getController($position){
        switch ($position) {
            case Land::LandA :
                return new LandA('A');
            case Land::LandB :
                return new LandB('B');
            case Land::LandC :
                return new LandC('C');
            case Land::LandD :
                return new LandD('D');
            case Land::LandE :
                return new LandE('E');
            default :
                return new LandStart('S');
        }
    }
}
class Land
{
    protected $landName; //島名

    const LandStart = 0;
    const LandGoal = 99;
    const LandA = 1;
    const LandB = 2;
    const LandC = 3;
    const LandD = 4;
    const LandE = 5;

    function __construct($landName)
    {
        $this->landName = $landName;
    }
    function getDice($arrInput, &$i): string
    {
        return $arrInput[$i];
    }
    function getCurrentTurnRecord($dice, $prefix): string
    {
        return $prefix . $this->landName . " " . $dice;
    }
    function getNextPosition($dice, &$gold): int
    {
        return 0;
    }
    function isGusu($dice): bool
    {
        return (($dice % 2) == 0) ? true : false;
    }
}
class LandStart extends Land
{
    function getNextPosition($dice, &$gold): int
    {
        switch ($dice) {
            case '1':
                return Land::LandA;
            case '2':
                return Land::LandB;
            case '6':
                return Land::LandC;
            default :
                return Land::LandStart;
        }
    }
}
class LandA extends Land
{
    function getNextPosition($dice, &$gold): int
    {
        switch ($dice) {
            case '3':
                return Land::LandB;
            case '4':
                return Land::LandC;
            default :
                $gold += 100;
                return Land::LandStart;
        }
    }
}
class LandB extends Land
{
    function getNextPosition($dice, &$gold): int
    {
        if (parent::isGusu($dice) === true) {
            return Land::LandE;
        }

        return Land::LandD;
    }
}
class LandC extends Land
{
    function getDice($arrInput, &$i): string
    {
        //サイコロ2回分の出目を取得する
        $dice = parent::getDice($arrInput, $i);
        $i++;
        $dice .= parent::getDice($arrInput, $i);
        return $dice;
    }
    function getNextPosition($dice, &$gold): int
    {
        //サイコロ2回分の出目を加算する
        $dice = (int)$dice[0] + (int)$dice[1];
        if (parent::isGusu($dice) === true) {
            $gold += 100;
            return Land::LandE;
        }
        $gold += 200;
        return Land::LandD;
    }
}
class LandD extends Land
{
    function getNextPosition($dice, &$gold): int
    {
        switch ($dice) {
            case '4':
            case '5':
                return Land::LandE;
            default :
                return Land::LandStart;
        }
    }
}
class LandE extends Land
{
    function getNextPosition($dice, &$gold): int
    {
        if ($dice == '6' && $gold >= 500) {
            return Land::LandGoal;
        }
        $gold += 100;
        return Land::LandC;
    }
}

File: backend/TreasureHunt/python/treasure_hunt.py
class Land:
    LandStart = 0
    LandGoal = 99
    LandA = 1
    LandB = 2
    LandC = 3
    LandD = 4
    LandE = 5

    def __init__(self, name: str):
        self.name = name

    def get_dice(self, arr_input: str, index: int):
        """ターンでのダイス目取得"""
        return arr_input[index]

    def get_next_position(self, dice: str, gold: int):
        """次の島への移動。返り値に次の位置とゴールドを返す"""
        return Land.LandStart, gold

    def get_turn_record(self, dice: str):
        return f"{self.name} {dice}"

    def is_even(self, dice: int):
        return dice % 2 == 0


class LandStart(Land):
    def get_next_position(self, dice: str, gold: int):
        if dice == '1':
            return Land.LandA, gold
        elif dice == '2':
            return Land.LandB, gold
        elif dice == '6':
            return Land.LandC, gold
        return Land.LandStart, gold


class LandA(Land):
    def get_next_position(self, dice: str, gold: int):
        if dice == '3':
            return Land.LandB, gold
        elif dice == '4':
            return Land.LandC, gold
        gold += 100
        return Land.LandStart, gold


class LandB(Land):
    def get_next_position(self, dice: str, gold: int):
        if self.is_even(int(dice)):
            return Land.LandE, gold
        return Land.LandD, gold


class LandC(Land):
    def get_dice(self, arr_input: str, index: int):
        if index + 1 < len(arr_input):
            # サイコロを2回分まとめて取得
            return arr_input[index] + arr_input[index + 1]
        else:
            # 最後のターン付近で1文字しか残っていない場合は2回目を同じ目として取得
            return arr_input[index] + arr_input[index]


    def get_next_position(self, dice: str, gold: int):
        total = int(dice[0]) + int(dice[1])
        if self.is_even(total):
            gold += 100
            return Land.LandE, gold
        gold += 200
        return Land.LandD, gold


class LandD(Land):
    def get_next_position(self, dice: str, gold: int):
        if dice in ('4', '5'):
            return Land.LandE, gold
        return Land.LandStart, gold


class LandE(Land):
    def get_next_position(self, dice: str, gold: int):
        if dice == '6' and gold >= 500:
            return Land.LandGoal, gold
        gold += 100
        return Land.LandC, gold


class LandFactory:
    @staticmethod
    def get_land(position: int):
        return {
            Land.LandA: LandA("A"),
            Land.LandB: LandB("B"),
            Land.LandC: LandC("C"),
            Land.LandD: LandD("D"),
            Land.LandE: LandE("E"),
        }.get(position, LandStart("S"))


class TreasureHunt:
    def __init__(self):
        self.record = ""
        self.gold = 0
        self.position = Land.LandStart
        self.turn = 0

    def execute(self, number: int, arr_input: str):
        index = 0
        while index < len(arr_input):
            self.turn += 1
            land = LandFactory.get_land(self.position)

            # ダイス目を取得
            dice = land.get_dice(arr_input, index)

            # 記録
            prefix = "" if not self.record else ", "
            self.record += prefix + land.get_turn_record(dice)

            # 次の位置とゴールドを取得
            self.position, self.gold = land.get_next_position(dice, self.gold)

            # ゴール判定
            if self.position == Land.LandGoal:
                self.record += ", G"
                break

            # LandC は2文字使うので index を調整
            index += 2 if isinstance(land, LandC) else 1

        return f'{number}, "{arr_input}", {self.turn}, {self.gold}, "{self.record}"'
